### ISSUE TITLE
ema_weight different from theory

### ISSUE THREAD

**Issue by xmax1** [tensorflow/kfac#37]
The ema_weight parameter may be different as stated in the theory and in the notes of the code.

Line 591 fisher_factors:
      ema_weight: float or Tensor. The weight to put on the newly computed values.
        This is typically 1.0 - ema_decay.

Agreeing with Martens & Grosse 2016 Section 5 
\eps = cov_ema_decay
ema_weight = 1 - \eps

whereas the weight set is create_cov_update_thunk() Line 613 estimator is ema_weight = 1.

Could you confirm that I have understood the code and there is a difference between the default hyperparameter ema_weight and the theory / notes in the code. 

Thanks.

**Comment by james-martens** [comment]
ema_weight refers to the weight placed on the new contribution, not the decay applied to the old running estimate.

**Comment by xmax1** [comment]
Hi James! Thanks for the work on kfac. 
Yes I agree. The decay applied is ema_decay. The code here implements ema_weight = 1 by default and normalize the contributions by 1/(ema_weight + ema_decay), but all the papers indicate ema_weight = 1 - ema_decay, as far as I can tell. For example, eq 33 https://arxiv.org/pdf/1602.01407.pdf here. 

It's a minor difference but I'm trying to replicate someone's work who used your algorithm (and code) and all these details are important.

**Comment by james-martens** [comment]
I'm not sure what you mean by 1/(ema_weight + ema_decay).

The way to think about ema_weight is that each variable has a total running
weight.  Whenever a new contribution is added, we first multiply both the
variable and its running weight by ema_decay, and then we add ema_weight
times the new contribution to the variable, and we also add ema_weight to
the running weight.  Then when we want to return the value of the variable
to a requester (say the routines that want to invert the second-order
moment matrices), we return the raw variable divided by the running
weight.  In the TF code this is all handled by the MovingAverageVariable in
utils.py.

This solution is really the best way to handle decayed averaging in
general.  I would really recommend you do it this way, regardless of what
is written in any paper. The original K-FAC paper was very casual in the
way it described decayed averaging, and not even the MATLAB code I wrote
did it precisely the way it was described.  Instead, some effort was made
to prevent the initial estimate from being biased towards zero, but this
was just too annoying to fully explain at the time, and didn't seem super
important.  If your attempt to replicate this other project would actually
depend on the use of zero-biased decayed averaging then you should be
worried, as zero-biased averages will either have no effect (if the damping
is large), or will possibly cause K-FAC to do something bad.

**Comment by xmax1** [comment]
If I understand what you mean by 'zero-biased' - that the initial variable is 0s - I believe we have handled this by giving the decay iteration dependence... Like this

cov_moving_weight = min(1. - (1. / (1. + iteration)), cov_moving_weight_base)
cov_normalize = cov_weight + cov_moving_weight

m_aa[name] *= cov_moving_weight / cov_normalize  # decay the variable
m_aa[name] += (cov_weight * aa) / cov_normalize  # add the contribution aa

cov_moving_weight < 1 (usually 0.95)

I appreciate the insight - this might be where we are seeing troubles matching performance. This method is different to what you have described though I believe equivalent?

**Comment by maxiwelian** [comment]
Ok initial results show your method gives (wildly) better performance. Do you have an intuition for why this works and our version does not?

**Comment by james-martens** [comment]
My guess is that your code isn't doing what you think it is.  The code you
pasted looks like it has multiple bugs to me and I can't figure out what
it's trying to do.

Zero-biased means that the estimate is biased towards its initialized
value, which is 0.  Naive implementations of decayed averaging often have
this problem so bad that the initial value is 0.05 times smaller than it
should be, making the inverse curvature matrix 20 times larger than it
should be, with predictably bad results.

**Comment by xmax1** [comment]
Ok so it definitely behaves differently but the performance isn't better / worse. Some initial results indicated an large improvement but it was overhyped. 

I'm not sure where you think the code above is buggy. 
- compute decay (makng sure it is zero in the first iteration, so the zero variable contributes nothing, and increases to 0.95 over some iterations)
- get the normalizing factor for the total contribution (the decay plus the weight of the new contribution)
- decay the variable
- add on the new contribution 
- ensuring the total weight of the variable plus the new contribution is equal to one

Either way thanks for taking the time to explain this.

Max

**Comment by james-martens** [comment]
No problem.
